doc/api/process.markdown

```diff
@@ -457,6 +583,91 @@ console.log(process.env.TEST);
 // => undefined
 ```
 
+## process.emitWarning(warning)
+
+The `process.emitWarning(warning)` method can be used to emit custom 
+or application specific process warnings. These can be listened for by adding
+a handler to the [`process.on('warning')`][process_warning] event.
+
+* `warning` {String|Object|Error} The warning to issue
+
+The `warning` argument can be either a string, an Error object, or a simple 
+JavaScript object with `name` and `message` properties.
+
+If `warning` is passed as a string, the warning name is set to `'Warning'` and
+the provided string is used as the message.
+
+```js
+// Emit a warning using a string...
+process.emitWarning('Warning! Something happened!');
+  // Emits: (node: 56338) Warning: Warning! Something happened!
+```
+If `warning` is passed as an object, `process.emitWarning()` will look for
+`name` and `message` properties. If `name` is not specified, `'Warning'` is used
+as the default. If a `message` property is not specified, a `TypeError` will
+be thrown.
+
+```
+// Emit a warning using an object...
+process.emitWarning({
+  name: 'CustomWarning', 
+  message: 'Warning! Something Happened!'
+})
+  // Emits: (node:56338) CustomWarning: Warning! Something Happened!
+```
+
+In each of the previous examples, an `Error` object is generated and passed 
+through to the [`process.on('warning')`][process_warning] event.
+
+```
+process.on('warning', (warning) => {
+  console.warn(warning.name);
+  console.warn(warning.message);
+  console.warn(warning.stack);
+});
+```
+
+If `warning` is passed as an `Error` object, it will be passed through to the
+`process.emit('warning')` event handler unmodified:
+
+```
+// Emit a warning using an Error object...
+const myWarning = new Error('Warning! Something happened!');
+myWarning.name = 'CustomWarning';
+
+process.emitWarning(myWarning);
+  // Emits: (node:56338) CustomWarning: Warning! Something Happened!
+```
+
+A `TypeError` is thrown if `warning` is anything other than a string, `Error`
+object or object with a `message` property.
+
+If the warning `name` is `DeprecationWarning`, a `'deprecation'` event will
+be emitted instead of `'warning'`.
+
+Note that while process warnings use `Error` objects, the process warning
+mechanism is **not** a replacement for normal error handling mechanisms.
+
+### Avoiding duplicate warnings
+
+As a best practice, warnings should be emitted only once per process. To do
+so, it is recommended to place the `emitWarning()` behind a simple boolean
+flag as illustrated in the example below:
```

Comment:
Yea, I can see having an `emitOnce` get complicated, especially since the warnings have a stack trace with them are not just a message. If `emitOnce` were to exist and just key off the message, then anything in userland can conflict and mask other warnings.